### PR TITLE
Read `--patch-module` arguments from 'Patch_Modules_argsFile' file

### PR DESCRIPTION
**Pull request by the1derer**
This Enables CheckerMain to read all `--patch-module` arguments in `${CHECKERFRAMEWORK}/checker/dist/annotatedJDK/jdk${JDK_VERSION}/Patch_Modules_argfile`

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 1]
```diff
@@ -27,7 +27,11 @@
  *   <li>add the {@code javac.jar} to the runtime classpath of the process that runs the Checker
  *       Framework.
  *   <li>add {@code jdk8.jar} to the compile time bootclasspath of the javac argument list passed to
- *       javac
+ *       javac while using Java 8
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Start the sentence with this information: "If using Java 8, ...".  Better, give this bullet point two sub-bullet points, one for Java 8 and one for Java 9+.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 2]
```diff
@@ -27,7 +27,11 @@
...
+ *   <li>patch JDK modules while using Java9+ using {@code --patch-module
+ *       &lt;module&gt;=&lt;annotatedModule&gt;} annotated modules are present in {@code
+ *       checker/dist/annotatedJDK/jdk*\/} directory. All {@code --patch-module} arguments are
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Say what the programmer should do.  This point starts saying to use `--patch-module`, but then says one should *not* use `--patch-module`, which is confusing.  Don't include implementation details, which is everything about `--patch-module`.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 3]
```diff
@@ -58,7 +62,10 @@ public static void main(String[] args) {
         System.exit(exitStatus);
     }
 
-    /** The path to the annotated jdk jar to use. */
+    /**
+     * The path to the annotated jdk jar to use in case of Java8 or directory containing
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: End sentences with a period.  State more clearly that this is one of two things; for example, change "or" to ", or to a".

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 4]
```diff
@@ -694,7 +731,7 @@ private static void assertFilesExist(final List<File> expectedFiles) {
                 missingAbsoluteFilenames.add(missingFile.getAbsolutePath());
             }
             throw new RuntimeException(
-                    "The following files could not be located: "
+                    "The following files/directory could not be located: "
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: plural/singular mismatch

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java [hunk 5]
```diff
@@ -34,8 +34,10 @@
     public static final String JAVAC_PATH_OPT = "-javacJar";
 
     /**
-     * Option name for specifying an alternative jdk.jar location. The accompanying value MUST be
-     * the path to the jar file (NOT the path to its encompassing directory)
+     * Option name for specifying an alternative jdk.jar location in case of Java8. The accompanying
+     * value MUST be the path to the jar file (NOT the path to its encompassing directory) In Case
```
Comment (mernst) [inline javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Put periods at ends of sentences.  Put "if" at beginning of sentences -- you don't want the reader to read an entire sentence only to learn at the end that it is irrelevant.  I would put the two sentences in different paragraphs to make the structure easier to read.

**Comment by the1derer** [comment]
@mernst 
Thanks for the feedback, I have made all the changes as suggested.

I can't understand why tests are failing but they all seem to fail for  [similar reasons](https://travis-ci.org/eisop/checker-framework/jobs/550729711#L3251) i.e
 ```
[ERROR] Failed to execute goal com.helger.maven:ph-javacc-maven-plugin:3.0.0:javacc (javacc) on project javaparser-core: Execution javacc of goal com.helger.maven:ph-javacc-maven-plugin:3.0.0:javacc failed: Plugin com.helger.maven:ph-javacc-maven-plugin:3.0.0 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven:maven-artifact:jar:3.0 from/to central (https://repo.maven.apache.org/maven2): Access denied to: https://repo.maven.apache.org/maven2/org/apache/maven/maven-artifact/3.0/maven-artifact-3.0.jar , ReasonPhrase:Forbidden. -> [Help 1]
```
Is this because of something I changed or any other reason?

I checked some recent updates to typetools/checker-framework and they are failing too due to [the similar reason](https://travis-ci.org/typetools/checker-framework/jobs/550546471#L1002).

Thanks.

**Comment by mernst** [comment]
Those problems are due to network failures.  They are not your fault.  You can restart the jobs repeatedly until they pass.
Yesterday I made changes to typetools/checker-framework that might resolve them.  Pulling that in might resolve the problems, but I have not yet been able to verify this.

**Comment by mernst** [comment]
@the1derer You should just restart the failing jobs in a build.  You restarted all the jobs, which has these problems:

 * It wastes resources.
 * If you restart a previously-passing job, it might fail.  Thus, you might never reach a state where all the jobs pass.

**Comment by the1derer** [comment]
@mernst
Thank you very much for your feedback. But I don't have permission to just restart one/all job that is why I have to close and reopen the PR for restarting the Travis Build which restarts all tasks.

Is there an option to start just one job(like I can do in my repositories' Travis Tests)?

Thanks.

**Comment by mernst** [comment]
I had already restarted all the jobs that needed to be restarted, before you closed and re-opened the pull request.  (You had visibility into that, I think.)  To get permissions to the eisop organization, you need to ask @wmdietl.

**Comment by the1derer** [comment]
@mernst
Thanks for restarting the tests.
All tests passing now. Ready for review.

**Comment by mernst** [comment]
@wmdietl Can you give a review (next week when you are not so busy) and merge when ready?  Thanks.

**Comment by wmdietl** [comment]
Yes, I will do that. Are all your review comments addressed?

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 2 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 1 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 3 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 4 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done

Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java (hunk 5 above)
Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Done

**Comment by mernst** [comment]
> Yes, I will do that. Are all your review comments addressed?

I believe so.

**Comment by the1derer** [comment]
@wmdietl 
I have addressed all the review comments given by @mernst.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 6]
```diff
@@ -432,6 +453,24 @@ protected void addMainToArgs(final List<String> args) {
             args.add(quote(concatenatePaths(ppOpts)));
         }
 
+        // Get CHECKERFRAMEWORK environment variable
+        final String CHECKERFRAMEWORK = System.getenv("CHECKERFRAMEWORK");
+
+        // addding "--patch-module" arguments to compiler Arguments
+        for (final String argLine : expandArgFiles(argListFiles)) {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Does this really need to go through all argfiles?
Isn't this substitution only meaningful for the `Patch_Modules_argfile` file?
Why not define your own simpler format for that file and here you write code that makes the relative paths absolute?

**Comment by the1derer** [comment]
@wmdietl Thank you very much for your feedback. Ready for review again.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 7]
```diff
@@ -432,6 +450,29 @@ protected void addMainToArgs(final List<String> args) {
             args.add(quote(concatenatePaths(ppOpts)));
         }
 
+        if (PluginUtil.getJreVersion() > 8) {
+            // Get Patch_Module_argfile
+            final File patchModuleArgsFile = new File(jdkJar, "Patch_Modules_argfile");
+
+            // Get ${CHECKERFRAMEWORK}/checker/dist path
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Shouldn't this be relative to `jdkJar` as well? It seems odd that the `JDK_PATH_OPT` flag would only impact the location of the `"Patch_Modules_argfile"` file, but not where the actual modules are.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 8]
```diff
@@ -432,6 +450,29 @@ protected void addMainToArgs(final List<String> args) {
...
+            final String checkerframeworkDistDirectoryPath = checkerJar.getParentFile().getPath();
+
+            // adding "--patch-module" arguments to compiler Arguments
+            try {
+                for (final String argLine : PluginUtil.readFile(patchModuleArgsFile)) {
+                    String[] separateFlag = argLine.trim().split("\\s+");
+                    args.add(separateFlag[0]);
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: How about you verify here that this is `--patch-module`? Or are there other valid options?

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 9]
```diff
@@ -24,13 +24,19 @@
...
- *   <li>parse and implement any special options used by the Checker Framework, e.g., using
- *       "shortnames" for annotation processors
- *   <li>pass all remaining command-line arguments to the real javac
+ *   <li>Use the annotated JDK.
+ *       <ul>
+ *         <li>If using Java 8: add {@code jdk8.jar} to the compile-time bootclasspath of the javac
+ *             argument list passed to javac.
+ *         <li>If using Java 9+: patch JDK modules using {@code --patch-module
+ *             <module>=<annotatedModule>}. Annotated modules are present in the {@code
+ *             checker/dist/annotatedJDK/jdk*\/} directory.
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Shouldn't this part depend on the `JDK_PATH_OPT` flag? `checker/dist` is just the default value for it. (Or maybe it works differently and you need to explain how it is supposed to work...)

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 10]
```diff
@@ -432,6 +450,29 @@ protected void addMainToArgs(final List<String> args) {
             args.add(quote(concatenatePaths(ppOpts)));
         }
 
+        if (PluginUtil.getJreVersion() > 8) {
+            // Get Patch_Module_argfile
+            final File patchModuleArgsFile = new File(jdkJar, "Patch_Modules_argfile");
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Can you describe the format of this file somewhere?

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 7 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: I have changed this to be relative to jdkJar.

**Comment by the1derer** [comment]
Here is the format of `Patch_Module_argfile`:
[Patch_Modules_argfile.txt](https://github.com/eisop/checker-framework/files/3389318/Patch_Modules_argfile.txt)

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 10 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Please see the comment below. Thanks.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 8 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Since we are going to generate `Patch_Module_argsfile`, I am not sure if we should check for its contents as it is only going to contain `--patch-module` options.

**Comment by the1derer** [comment]
@wmdietl Thanks for the feedback. PR is ready for review again.
Thanks.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 8 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: I think it would be good to describe your assumptions about the `Patch_Module_argsfile` file and the content of that directory somewhere. The tool that will auto-generate that file will need to rely on that same specification. For example, the substitution of `CURRENT_JDK_FOLDER` should be explained somewhere.
Also, that file could be generated manually, so a user needs to know what the format is.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 10 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: In the example file that you sent it looks like you patched every single module. I think it would make sense to limit that to only the modules that actually contain useful information. That is, if you didn't annotate a particular module, don't patch that module in - it will not have an effect and we don't need to distribute that module.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 11]
```diff
@@ -58,7 +65,10 @@ public static void main(String[] args) {
         System.exit(exitStatus);
     }
 
-    /** The path to the annotated jdk jar to use. */
+    /**
+     * In the case of Java 8, the path to the annotated jdk.jar. In the case of Java 9+, the path to
+     * a directory containing annotated-jdk modules.
+     */
     protected final File jdkJar;
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: I suggest that you rename this `jdkPath` because the current name is not accurate for JDK 9+.

**Comment by mernst** [comment]
@the1derer I assigned you because of my comments two days ago.
@wmdietl remains assigned as well.

**Comment by the1derer** [comment]
@wmdietl 
I will first make improve according to the latest suggestions given by @mernst, as I need to discuss some of the aspects of your comments with @mernst.
Thanks.

**Comment by wmdietl** [comment]
Sure, just let me know when I should review this again.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 11 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 12]
```diff
@@ -97,6 +106,7 @@ public CheckerMain(final File checkerJar, final List<String> args) {
 
         this.checkerJar = checkerJar;
         final File searchPath = checkerJar.getParentFile();
+        final File annotatedJDKSearchPath = new File(checkerJar.getParentFile(), "annotatedJDK");
```
[the remark this answers is not in the document]
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 13]
```diff
@@ -432,6 +453,24 @@ protected void addMainToArgs(final List<String> args) {
             args.add(quote(concatenatePaths(ppOpts)));
         }
 
+        // Get CHECKERFRAMEWORK environment variable
+        final String CHECKERFRAMEWORK = System.getenv("CHECKERFRAMEWORK");
+
+        // addding "--patch-module" arguments to compiler Arguments
```
[the remark this answers is not in the document]
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done.

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java [hunk 14]
```diff
@@ -34,8 +34,13 @@
...
 
     /**
-     * Option name for specifying an alternative jdk.jar location. The accompanying value MUST be
-     * the path to the jar file (NOT the path to its encompassing directory)
+     * Option name for specifying an alternative location for the annotated JDK.
+     *
+     * <p>In case of Java 8, an alternative jdk.jar location. The accompanying value MUST be the
+     * path to the jar file (NOT the path to its encompassing directory).
+     *
+     * <p>In case of Java 9+. a directory containing alternate &lt;module&gt;.jar which will patch
```
[the remark this answers is not in the document]
Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Done.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 15]
```diff
@@ -26,8 +26,14 @@
  * <ul>
  *   <li>add the {@code javac.jar} to the runtime classpath of the process that runs the Checker
  *       Framework.
- *   <li>add {@code jdk8.jar} to the compile time bootclasspath of the javac argument list passed to
- *       javac
+ *   <li>Use the annotated JDK.
+ *       <ul>
+ *         <li>If using Java 8: add {@code jdk8.jar} to the compile-time bootclasspath of the javac
+ *             argument list passed to javac.
+ *         <li>If using Java 9+: patch JDK modules using {@code --patch-module
```
[the remark this answers is not in the document]
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: I will make the comments more precise after (annotated)JDK12 is uploaded somewhere.

**Comment by the1derer** [comment]
@mernst 
Ready for review.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 10 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: One approach is to provide every module, even those that don't contain any changes.
The list of provided modules would only change when the Checker Framework supports a new JDK.

Another approach is to determine which modules contain annotations (maybe by grepping for `import org.checkerframework`?), and only provide those modules.
The list of patched modules could change for any Checker Framework release -- whenever the Checker Framework developers annotate more of the JDK.

The former seems simpler, and requires less-frequent work by Checker Framework users.
What are the key benefits of the latter that motivate it?  (Efficiency, or the size of the distribution, or something else?)

**Comment by mernst** [comment]
@the1derer could you please address the continuous integration failures that you can?
There are 7 CI jobs, and 5 of them are failing.  They are failing for 3 or 4 completely different reasons.  Can you please investigate each of them?  I think several can be resolved independently of the others.
It is a bad idea to ignore some failures, because they might be indirectly causing others.

**Comment by the1derer** [comment]
@mernst All tests that are passing on base-branch are passing now.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 16]
```diff
@@ -58,9 +66,15 @@ public static void main(String[] args) {
         System.exit(exitStatus);
     }
 
-    /** The path to the annotated jdk jar to use. */
+    /** In case of Java 8, the path to the annotated jdk.jar. Otherwise {@code null}. */
     protected final File jdkJar;
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: To improve the documentation, you could add `@Nullable` here and on the new field type.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 17]
```diff
@@ -110,11 +124,24 @@ public CheckerMain(final File checkerJar, final List<String> args) {
...
+        if (PluginUtil.getJreVersion() > 8) {
+            final File annotatedJDKSearchPath = new File(searchPath, "annotatedJDK");
+            final int jreVersion = PluginUtil.getJreVersion();
+            final String jdkVersionFolderName = "jdk" + jreVersion;
+            final File jdkVersionModuleLocation =
+                    new File(annotatedJDKSearchPath, jdkVersionFolderName);
+            this.jdkModulesPath =
+                    extractFileArg(PluginUtil.JDK_PATH_OPT, jdkVersionModuleLocation, args);
+            this.jdkJar = null;
+            this.compilationBootclasspath = null;
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Please document the changed meaning of `compilationBootclasspath` and mark that field as `@Nullable`.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 18]
```diff
@@ -432,6 +467,28 @@ protected void addMainToArgs(final List<String> args) {
...
 
+        if (PluginUtil.getJreVersion() > 8) {
+            // Get Patch_Module_argfile
+            final File patchModuleArgsFile = new File(jdkModulesPath, "Patch_Modules_argfile");
+
+            // Get currentJdkFolder
+            final String currentJdkFolder = jdkModulesPath.getPath();
+
+            // adding "--patch-module" arguments to compiler arguments
+            try {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Can you change the `try` block to cover the smallest possible region? Isn't `PluginUtil.readFile` the only call that can cause an exception? Limiting the scope will make the error situations easier to understand.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 8 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Again, could you add such documentation somewhere? You now added a bit in `PluginUtil` below, but that does not seem like the best place for this information. Even a few inline comments here would be helpful.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java [hunk 19]
```diff
@@ -694,7 +751,7 @@ private static void assertFilesExist(final List<File> expectedFiles) {
                 missingAbsoluteFilenames.add(missingFile.getAbsolutePath());
             }
             throw new RuntimeException(
-                    "The following files could not be located: "
+                    "The following files and/or directory could not be located: "
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: `files` is plural, but `directory` is singular.

File: javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java [hunk 20]
```diff
@@ -624,7 +624,10 @@ public static boolean hasElementValue(AnnotationMirror anno, CharSequence elemen
                 return expectedType.cast(val.getValue());
             }
         }
-        throw new BugInCF("No element with name \'" + elementName + "\' in annotation " + anno);
+        throw new BugInCF(
```
Comment (wmdietl) [inline javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java]: This looks like an unrelated change that should go directly to typetools.

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java [hunk 21]
```diff
@@ -34,11 +34,31 @@
...
 
     /**
-     * Option name for specifying an alternative jdk.jar location. The accompanying value MUST be
-     * the path to the jar file (NOT the path to its encompassing directory)
+     * Option name for specifying an alternative location for the annotated JDK.
+     *
+     * <p>In case of Java 8, an alternative jdk.jar location. The accompanying value MUST be the
+     * path to the jar file (NOT the path to its encompassing directory).
+     *
+     * <p>In case of Java 9+, a directory containing annotated &lt;module&gt;.jar. The supplied
```
Comment (wmdietl) [inline javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: How about:
```
     * <p>In case of Java 9+, a directory containing annotated JDK modules. The supplied
     * directory MUST contain a {@code Patch_Modules_argfile} file containing {@code --patch-module
     * <jdkModule>=<annotatedModule>} directives, one per line. The string `${CURRENT_JDK_FOLDER}` will
     * be replaced with this directory. The directory must also contain the refrenced &lt;module&gt;.jar files.
```

However, this seems like an odd place to document the file format. So move this somewhere else and refer to it here.

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java [hunk 22]
```diff
@@ -34,11 +34,31 @@
...
+    static {
+        if (PluginUtil.getJreVersion() == 8) {
+            JDK_PATH_OPT = "-jdkJar";
+        } else {
+            JDK_PATH_OPT = "-jdkModulesPath";
+        }
+    }
 
+    /**
+     * Takes list of fileNames and creates list of files.
```
Comment (wmdietl) [inline javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: `Convert a list of file names into a list of files.`
`@param fileNames list of file names`

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 10 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Not sure whether this is for Shubham or me and nearly missed it...

As file `Patch_Modules_argfile` is generated, there shouldn't be an impact on users - they either use `CheckerMain` or import that argfile.

I assume we don't have annotations for many of the JDK modules, so distributing all of them seems wasteful and misleading. Maybe that assumption is wrong. I would think most of the current annotations ended up in `java.base.jmod` and some in management and xml.
@the1derer Can you already determine in which modules the current annotations ended up in?

Including only the ones that are `@AnnotatedFor` for one of the type systems would make the most sense to me, but if most modules are (partially) annotated I'm fine with simply listing all.

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java (hunk 21 above)
Comment (wmdietl) [inline javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Instead of `.jar` should this be `.jmod`?

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 10 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: @wmdietl Thanks for the clarification. Wasteful in terms of what?  Run time, bandwidth, or something else?  What is the impact for users?

@wmdietl @the1derer How is `Patch_Modules_argfile` generated?  My main concern was extra steps (and the ability to forget them!) when a CF developer adds new JDK annotations.  I was willing to make the distribution larger in order to avoid those steps and mistakes.  But I don't have concrete numbers to base that decision on.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 19 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: There can only be one directory missing like `jdk11`/`jdk12` but there can be more than one files that could be missing.

File: javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java (hunk 20 above)
Comment (mernst) [reply javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java]: It is already in typetools.  This diff shows up here because eisop is behind typetools (and because that typetools change was cherry-picked into this branch).
I have made a pull request: https://github.com/typetools/checker-framework/pull/2645
If you could complete that (I don't think my input is necessary), that would be great.
Once the changes in eisop that are not related to JDK 9 are pull-requested into typetools, we can avoid making painful merges like that one.
Thanks!

Comment (wmdietl) [reply javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java]: master should now be in sync with typetools.
@the1derer Can you merge in master and make sure this change disappears?

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java (hunk 21 above)
Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: https://docs.google.com/document/d/1mcOJUrYzHDMp2MfPztOmOaaSFjOWZquUN1fO_LCmXuo/edit?disco=AAAACPnypBE

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java (hunk 22 above)
Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Done.

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java (hunk 21 above)
Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Done.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 18 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 17 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 16 above)
Comment (the1derer) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: Done.

File: javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java (hunk 21 above)
Comment (wmdietl) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: Can you be a bit more specific? That document states "We will have replacement .jmod files, e.g. an annotated java.base.jmod for most of our existing annotations".
So why is this directory containing `.jar` files instead of `.jmod` files?

Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: We can use both `.jmod` and `.jar` files for `--patch-module` argument. Since it is easier to create `.jar` file I went with this option.
I also discussed this with @mernst when I was first starting and he agrees with me.
Is there any advantage to use `.jmod`?

Comment (mernst) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: More precisely, I didn't know a reason to object to .jar files.
@wmdietl are .jmod files better?

Comment (wmdietl) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: The normal OpenJDK build creates .jmod files. If you follow the [normal build instructions](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/master/doc/building.md), you get all the .jmod files in `build/linux-x86_64-server-release/images/jmods/` or some similar directory.
So instead of asking why .jmod files are better, why should we add additional build steps and create separate .jar files?

Comment (mernst) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: @the1derer , can you explain your comment that it's easier to create the `.jar` files?
Sticking to the normal build process would be good, if possible.

Comment (the1derer) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: @mernst @wmdietl Apologies for the late reply. I was checking how to use JMOD files.

I agree sticking to the normal build process would be good, but after compiling (annotated)JDK11 I tried using JMOD files in `build/linux-x86_64-server-release/images/jmods/` via `--patch-module java.base=java.base.jmod` but it didn't seem to work(ie. it is still issuing,`warning: You do not seem to be using the distributed annotated JDK.`) on the other hand, using `build/linux-x86_64-server-release/jdk/modules`(from which I created my JAR files) seem to work fine.
I am not able to understand why this problem is occurring.

Comment (mernst) [reply javacutil/src/main/java/org/checkerframework/javacutil/PluginUtil.java]: In other words:  --patch-module works with jar files but not with jmod files.

**Comment by the1derer** [comment]
@wmdietl Ready for review.

File: framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java (hunk 19 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/CheckerMain.java]: That assumes a certain use of that method, which is not documented in the javadoc.
I think in general it reads nicer to be consistent about singular/plural. That is, you could check the length of `missingAbsoluteFilenames` and if it's one use singular and otherwise plural. Or consistently use singular/plural, regardless of how many files or directories are not found.

**Comment by wmdietl** [comment]
@the1derer Now that the JDK 11 changes are merged into typetools/master, can you please open a PR with these changes there?

**Comment by the1derer** [comment]
> @the1derer Now that the JDK 11 changes are merged into typetools/master, can you please open a PR with these changes there?

@wmdietl  Done.
Please review at typetools/checker-framework#2773
